### PR TITLE
server: automatically retry CI testing failures

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Leverage `--rerun-fails=3` for `make test-server` and `make test-server-race`. Tests that fail on retry will still get reported in the summary and will now also ping devs in https://community.mattermost.com/core/channels/developers. Note that `gotestsum` defaults to `--rerun-fails-max-failures=10` and won't retry the whole suite multiple times in the case of a major regression.

Flaky tests still get reported as failures, but won't fail the actual build. A notification https://community.mattermost.com/core/channels/developers will also alert devs to the need to take action:

image

#### Ticket Link
None.

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: .github/workflows/server-ci-template.yml [hunk 1]
```diff
@@ -9,289 +9,11 @@ env:
   go-version: "1.19.5"
 
 jobs:
-  check-mocks:
```
Comment (lieut-data) [inline .github/workflows/server-ci-template.yml:12]: These removals are just to make iterating on the CI changes faster, and will be reverted before opening up the PR for final review.

File: .github/workflows/server-test-template.yml [hunk 2]
```diff
@@ -19,7 +19,7 @@ env:
 jobs:
   test:
     name: ${{ inputs.name }}
-    runs-on: ubuntu-latest-8-cores
+    runs-on: ubuntu-22.04
```
Comment (lieut-data) [inline .github/workflows/server-test-template.yml:22]: Temporary for testing.

File: server/Makefile [hunk 3]
```diff
@@ -151,9 +151,9 @@ DIST_PATH_OSX_ARM64=$(DIST_ROOT)/osx_arm64/mattermost
 DIST_PATH_WIN=$(DIST_ROOT)/windows/mattermost
 
 # Packages lists
-TE_PACKAGES=$(shell $(GO) list ./... | grep -vE 'server/v8/cmd/mmctl')
-SUITE_PACKAGES=$(shell $(GO) list ./...| grep -vE 'server/v8/cmd/mmctl')
-MMCTL_PACKAGES=$(shell $(GO) list ./... | grep -E 'server/v8/cmd/mmctl')
+TE_PACKAGES=$(shell $(GO) list ./... | grep -vE 'server/v8/cmd/mmctl' | grep 'server/v8/test')
```
Comment (lieut-data) [inline server/Makefile:154]: Temporary for testing.

File: server/Makefile [hunk 4]
```diff
@@ -189,7 +189,7 @@ else
 	IGNORE:=$(shell rm -f channels/imports/imports.go)
 endif
 
-EE_PACKAGES=$(shell $(GO) list $(BUILD_ENTERPRISE_DIR)/...)
+EE_PACKAGES=
```
Comment (lieut-data) [inline server/Makefile:192]: Temporary for testing.

File: server/test/main_test.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package main
```
Comment (lieut-data) [inline server/test/main_test.go:1]: Temporary for testing.

File: .github/workflows/server-test-template.yml [hunk 6]
```diff
@@ -68,10 +68,15 @@ jobs:
...
         if: success() || failure() # always run even if the previous step fails
         with:
           report_paths: server/report.xml
           check_name: ${{ inputs.name }} (Results)
-          job_name: ${{ inputs.name }}
+          require_tests: true
+      - name: Report retried tests via webhook
+        if: ${{ steps.report.outputs.retried > 0 }}
+        run: |
+          curl -d "{\"text\":\"@devs: ⚠️ One or more flaky tests retried in [github.com/mattermost/mattermost: ${{ inputs.name}}](${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }})\"}" -H "Content-Type: application/json" -X POST https://community.mattermost.com/hooks/kehmczwh93bjidunmchbf4xn4c
```
Comment (lieut-data) [inline .github/workflows/server-test-template.yml:82]: Temporary webhook that will be replaced with a secret before merging.

**Comment by lieut-data** [comment]
@phoinixgrr, could I ask for your help reviewing the GitHub Actions changes in question?

File: server/Makefile [hunk 7]
```diff
@@ -470,7 +470,7 @@ test-server-race: test-server-pre
 ifeq ($(IS_CI),true)
 	GOMAXPROCS=4 $(GOBIN)/gotestsum --packages="$(TE_PACKAGES) $(EE_PACKAGES)" -- -race $(GOFLAGS) -timeout=90m
```
Comment (agnivade) [inline server/Makefile:471]: Don't we need to add it here?

**Review by larkox — APPROVED** [review]

**Review by phoinixgrr — APPROVED** [review]
🎖️  No objections merging this.

**Review by agnivade — APPROVED** [review]
Thanks, looking forward to this!

**Pull request merged**